File: test/aidbox_sdk/snapshots/python/hl7_fhir_r4_core/MedicinalProduct.py
from typing import Optional, List
from dataclasses import dataclass, field
from base import CodeableConcept
from base import MarketingStatus
from base import Coding
from base import DomainResource
from base import Reference
from base import Identifier
from base import BackboneElement

@dataclass(kw_only=True)
class MedicinalProductManufacturingBusinessOperation(BackboneElement):
    regulator: Optional[Reference] = None
    manufacturer: Optional[List[Reference]] = field(default_factory=list)
    effective_date: Optional[str] = None
    operation_type: Optional[CodeableConcept] = None
    confidentiality_indicator: Optional[CodeableConcept] = None
    authorisation_reference_number: Optional[Identifier] = None

@dataclass(kw_only=True)
class MedicinalProductNameNamePart(BackboneElement):
    part: str
    type: Coding

@dataclass(kw_only=True)
class MedicinalProductNameCountryLanguage(BackboneElement):
    country: CodeableConcept
    language: CodeableConcept
    jurisdiction: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class MedicinalProductName(BackboneElement):
    name_part: Optional[List[MedicinalProductNameNamePart]] = field(default_factory=list)
    product_name: str
    country_language: Optional[List[MedicinalProductNameCountryLanguage]] = field(default_factory=list)

@dataclass(kw_only=True)
class MedicinalProductSpecialDesignation(BackboneElement):
    date: Optional[str] = None
    species: Optional[CodeableConcept] = None
    type: Optional[CodeableConcept] = None
    intended_use: Optional[CodeableConcept] = None
    status: Optional[CodeableConcept] = None
    identifier: Optional[List[Identifier]] = field(default_factory=list)
    indication_codeable_concept: Optional[CodeableConcept] = None
    indication_reference: Optional[Reference] = None

@dataclass(kw_only=True)
class MedicinalProduct(DomainResource):
    additional_monitoring_indicator: Optional[CodeableConcept] = None
    manufacturing_business_operation: Optional[List[MedicinalProductManufacturingBusinessOperation]] = field(default_factory=list)
    combined_pharmaceutical_dose_form: Optional[CodeableConcept] = None
    clinical_trial: Optional[List[Reference]] = field(default_factory=list)
    product_classification: Optional[List[CodeableConcept]] = field(default_factory=list)
    name: list[MedicinalProductName] = field(default_factory=list)
    master_file: Optional[List[Reference]] = field(default_factory=list)
    pharmaceutical_product: Optional[List[Reference]] = field(default_factory=list)
    type: Optional[CodeableConcept] = None
    marketing_status: Optional[List[MarketingStatus]] = field(default_factory=list)
    special_measures: Optional[List[str]] = field(default_factory=list)
    special_designation: Optional[List[MedicinalProductSpecialDesignation]] = field(default_factory=list)
    packaged_medicinal_product: Optional[List[Reference]] = field(default_factory=list)
    identifier: Optional[List[Identifier]] = field(default_factory=list)
    cross_reference: Optional[List[Identifier]] = field(default_factory=list)
    attached_document: Optional[List[Reference]] = field(default_factory=list)
    domain: Optional[Coding] = None
    legal_status_of_supply: Optional[CodeableConcept] = None
    paediatric_use_indicator: Optional[CodeableConcept] = None
    contact: Optional[List[Reference]] = field(default_factory=list)

File: test/aidbox_sdk/snapshots/dotnet/hl7-fhir-r4-core/MedicinalProduct.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class MedicinalProduct : DomainResource, IResource
{
    public Base.CodeableConcept? AdditionalMonitoringIndicator { get; set; }
    public MedicinalProduct_ManufacturingBusinessOperation[]? ManufacturingBusinessOperation { get; set; }
    public Base.CodeableConcept? CombinedPharmaceuticalDoseForm { get; set; }
    public Base.ResourceReference[]? ClinicalTrial { get; set; }
    public Base.CodeableConcept[]? ProductClassification { get; set; }
    public required MedicinalProduct_Name[] Name { get; set; }
    public Base.ResourceReference[]? MasterFile { get; set; }
    public Base.ResourceReference[]? PharmaceuticalProduct { get; set; }
    public Base.CodeableConcept? Type { get; set; }
    public Base.MarketingStatus[]? MarketingStatus { get; set; }
    public string[]? SpecialMeasures { get; set; }
    public MedicinalProduct_SpecialDesignation[]? SpecialDesignation { get; set; }
    public Base.ResourceReference[]? PackagedMedicinalProduct { get; set; }
    public Base.Identifier[]? Identifier { get; set; }
    public Base.Identifier[]? CrossReference { get; set; }
    public Base.ResourceReference[]? AttachedDocument { get; set; }
    public Base.Coding? Domain { get; set; }
    public Base.CodeableConcept? LegalStatusOfSupply { get; set; }
    public Base.CodeableConcept? PaediatricUseIndicator { get; set; }
    public Base.ResourceReference[]? Contact { get; set; }

    public class MedicinalProduct_ManufacturingBusinessOperation : BackboneElement
    {
        public Base.ResourceReference? Regulator { get; set; }
        public Base.ResourceReference[]? Manufacturer { get; set; }
        public string? EffectiveDate { get; set; }
        public Base.CodeableConcept? OperationType { get; set; }
        public Base.CodeableConcept? ConfidentialityIndicator { get; set; }
        public Base.Identifier? AuthorisationReferenceNumber { get; set; }
    }

    public class MedicinalProduct_Name_NamePart : BackboneElement
    {
        public required string Part { get; set; }
        public required Base.Coding Type { get; set; }
    }

    public class MedicinalProduct_Name_CountryLanguage : BackboneElement
    {
        public required Base.CodeableConcept Country { get; set; }
        public required Base.CodeableConcept Language { get; set; }
        public Base.CodeableConcept? Jurisdiction { get; set; }
    }

    public class MedicinalProduct_Name : BackboneElement
    {
        public MedicinalProduct_Name_NamePart[]? NamePart { get; set; }
        public required string ProductName { get; set; }
        public MedicinalProduct_Name_CountryLanguage[]? CountryLanguage { get; set; }
    }

    public class MedicinalProduct_SpecialDesignation : BackboneElement
    {
        public string? Date { get; set; }
        public object? Indication    
        {
            get
            {
                if (IndicationCodeableConcept is not null)
                {
                    return IndicationCodeableConcept;
                }
        
                if (IndicationReference is not null)
                {
                    return IndicationReference;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    IndicationCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    IndicationReference = (Base.ResourceReference)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.CodeableConcept? Species { get; set; }
        public Base.CodeableConcept? Type { get; set; }
        public Base.CodeableConcept? IntendedUse { get; set; }
        public Base.CodeableConcept? Status { get; set; }
        public Base.Identifier[]? Identifier { get; set; }
        public Base.CodeableConcept? IndicationCodeableConcept { get; set; }
        public Base.ResourceReference? IndicationReference { get; set; }
    }
}